### PR TITLE
DM-47769: Add pagination support and more admin routes

### PR DESCRIPTION
**Pull request by rra**
Convert the `/jobs` route to support pagination. Change `count` to `limit` and `after` to `since` to match Gafaelfawr and be more precise (`after` could be interpreted as a generic sorting term, while `since` is always about time), since we're not following the current IVOA UWS API anyway.

Convert the existing admin jobs route to pagination and add multiple new admin routes, allowing the listing of users and jobs for all services and, in the later case, for all users as well.

Refactor the additional limiting criteria for searches into a dependency so that it can be shared by the various routes.

### CODE REVIEW

File: tests/handlers/service_test.py [hunk 1]
```diff
@@ -370,3 +375,93 @@ async def test_errors(client: AsyncClient) -> None:
...
+            new_creation = datetime.now(tz=UTC) - timedelta(minutes=5)
+            job.creation_time = datetime_to_db(new_creation)
+    r = await client.get(
+        "/wobbly/jobs",
+        params={"since": (now - timedelta(seconds=1)).isoformat()},
+        headers=headers,
+    )
+    assert r.status_code == 200
+    since_expected = [*expected[:8], expected[9]]
+    assert [j["parameters"]["id"] for j in r.json()] == since_expected
```
Comment (stvoutsin) [inline tests/handlers/service_test.py:467]: Some edge case tests (perhaps future work):

Edge cases:

- Query with limit=0 or limit=-1 (or even non-numeric) limit
- Pagination with limit larger than the total number of jobs
- Query without a phase parameter but with other unrelated parameters
- Creation time filtering: No jobs match the since parameter
- Test behaviour when no jobs exist

I think at first glance these would mostly be fine but perhaps good for regression testing

File: src/wobbly/storage.py [hunk 2]
```diff
@@ -169,48 +177,43 @@ async def list_expired(self) -> list[Job]:
 
     async def list_jobs(
```
Comment (stvoutsin) [inline src/wobbly/storage.py:178]: Could the`@retry_async_transaction` decorator be useful here as well?

File: src/wobbly/dependencies/search.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+from fastapi import Query
+from safir.pydantic import UtcDatetime
+from vo_models.uws.types import ExecutionPhase
+
+from ..models import JobCursor, JobSearch
+
+__all__ = ["job_search_dependency"]
+
+
+async def job_search_dependency(
```
Comment (stvoutsin) [inline src/wobbly/dependencies/search.py:14]: I'm wondering if we're missing the "LAST" query param for UWS jobs?

> /{jobs}?LAST=100 The server should list only the given number (an integer greater than 0) of most recent jobs ordered by descending creationTimes.

**Review by stvoutsin — APPROVED** [review]
This all looks great! I couldn't see any obvious issues, so only a few small suggestions

File: src/wobbly/dependencies/search.py (hunk 3 above)
Comment (rra) [reply src/wobbly/dependencies/search.py:14]: This can be implemented with limit. I had originally been sticking to the UWS parameters, but they're a little weird compared to the rest of our services so I decided to align this with the pagination documentation in Safir and with Gafaelfawr (which is where that came from), and will convert from the UWS names to these names in the Wobbly client code that will go into Safir.

File: src/wobbly/storage.py (hunk 2 above)
Comment (rra) [reply src/wobbly/storage.py:178]: I don't *think* that a pure read transaction can fail and thus shouldn't need to be retried. The write transactions need this because of the isolation level, since they do a read and then make changes based on that read and therefore need to fail if the data they read was out of date at the time of the commit. But that scenario shouldn't apply here, I think.

File: tests/handlers/service_test.py (hunk 1 above)
Comment (rra) [reply tests/handlers/service_test.py:467]: Thanks, I've added several of these. I tend not to do full black box testing and instead do testing somewhat informed by the implementation so I don't test every combinatoric when the implementation is unrelated just to save from generating endless test cases, so I think for example the variations of phase are covered. I also tend to avoid testing basic Pydantic functionality (like non-numeric limit). But limit 0 and -1 are great edge cases, and I missed several of the pagination cases that you mentioned as well as several of the empty result tests. Those have now been added.

File: src/wobbly/dependencies/search.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            title="Pagination cursor",
+            description="Cursor used when moving between pages of results",
+        ),
+    ] = None,
+    limit: Annotated[
+        int | None,
+        Query(
+            title="Number of jobs",
+            description="Return at most the given number of jobs",
+        ),
```
[the remark this answers is not in the document]
Comment (rra) [reply src/wobbly/dependencies/search.py:43]: `ge=1` is a great idea, added. Since we allow not setting the limit at all, I decided that having an upper limit didn't make sense. We can cap the limit and force pagination later if that ends up being a problem, but since the common case will be passing along a UWS query, and UWS currently doesn't support pagination, we're stuck with the memory usage hit in the service anyway and I'm not sure it's worth restricting it here.

File: tests/handlers/service_test.py [hunk 5]
```diff
@@ -370,3 +375,93 @@ async def test_errors(client: AsyncClient) -> None:
             **cast(dict[str, Any], kwargs),
         )
         assert r.status_code == 404
+
+
+@pytest.mark.asyncio
+async def test_pagination(client: AsyncClient) -> None:
```
[the remark this answers is not in the document]
Comment (rra) [reply tests/handlers/service_test.py:382]: Thanks, good call, I was being lazy.
